Add killed and deleted jobs to lists of job states and final states

meliache opened the pull request:
Dear Dirac Developers,

For some reason, the job states `DELETED` and `KILLED` are defined in the `JobStatus.py`, but they are not included into the lists `JOB_STATES` and `FINAL_JOB_STATES`. I don't know why that is, in this PR I added them, view this as a suggestion.

I am just a Dirac user at Belle II and wanted to find out possible job states for script that I am writing and so I checked the Dirac code, where I found this file. These lists are used in `JobDB.py` and `JobStateUpdateHandler.py`, If for some reason the `DELETED` and `KILLED` jobs are to be ignored in those files, I would still keep a list of all possible job states and just make sure at the place where they are used that deleted and killed jobs are ignored. Or maybe make another subset and use that. As a user, I would like to be able to import JobStatus and use the list of all possible job statuses, so I think it is important to keep a list with all known job statuses, instead of having them just as global variables.

I am not sure if this PR is to the correct branch, since the `JobStatus.py` also exists e.g. v7r1, I just used the default branch. Maybe I should have made a github issue or write and email, but I thought it might be easier to just do a PR if I have a code change idea.

By the way, I also think one could use a python set for the job statuses (`{...}` instead of `[...]`), to avoid accidental duplicates and make it easy to to unions and differences of different job state sets. Or maybe the most elegent is an enum type, which I have seen in other code bases used for job states. But Ithose ideas are irrelevant to this PR.

Best regards,
Michael Eliachevitch ([meliache@uni-bonn.de](mailto:meliache@uni-bonn.de))

phicharp commented:
These statuses are not final states, as Killed jobs should move to Failed after a while, while Deleted jobs are soon... deleted, so in some sense, yes it is a final state ;-)
I would agree to add them to the list of states, you are right but not to the list of final states...

meliache commented:
@phicharp 
> These statuses are not final states, as Killed jobs should move to Failed after a while, while Deleted jobs are soon... deleted, so in some sense, yes it is a final state ;-)
> I would agree to add them to the list of states, you are right but not to the list of final states...

Thanks for the quick reply. I just removed the changes of the final states in b450301c05aeb9e8609f888f59764d3fe03e366e.

Also a related question, which is probably due to my naivity, as I am new to Dirac, but won't `Completed` jobs also eventually transition to `Done`? I remember reading somewhere that it means that the output files have yet to be moved to the correct site, but I am not sure. Well, this last step is maybe not  really part of the job, so maybe from the point of view of the job it is still final?

phicharp commented:
You are right, `Completed` is strictly speaking not a final state of the job... I left it here as it was before I created this file... I didn't dare changing that as it may have side effect in case somewhere the code assumed that this state is in `FINAL_JOB_STATES`. I can have a deeper look at where this is used...

meliache commented:
Thanks for the information, at least I learned something from this PR. Now I tend more towards thinking a mail with a questions would also have sufficed. It seems the `checkErrors` test fails, but I don't know why exactly. Maybe I should leave that to the experts and if you want you can close the PR.

atsareg commented:
This PR should rather target the rel-v7r1 release branch. I will include this changes in the coming v7r1 release and once done I will close this PR

fstagni commented:
Cherry-picked in #4556 so this can be closed. Thanks @meliache
